r/src/safe-call-into-r.h

```diff
@@ -20,18 +20,25 @@
 
 #include "./arrow_types.h"
 
+#include <arrow/io/interfaces.h>
 #include <arrow/util/future.h>
 #include <arrow/util/thread_pool.h>
 
 #include <functional>
 #include <thread>
 
 // Unwind protection was added in R 3.5 and some calls here use it
-// and crash R in older versions (ARROW-16201). We use this define
-// to make sure we don't crash on R 3.4 and lower.
+// and crash R in older versions (ARROW-16201). Crashes also occur
+// on 32-bit R builds on R 3.6 and lower.
+static inline bool CanSafeCallIntoR() {
 #if defined(HAS_UNWIND_PROTECT)
-#define HAS_SAFE_CALL_INTO_R
+  cpp11::function on_old_windows_fun = cpp11::package("arrow")["on_old_windows"];
+  bool on_old_windows = on_old_windows_fun();
```

Comment:
Just out of curiosity, is it costly to call into the R interpreter like this everytime `CanSafeCallIntoR`? Should perhaps the result be cached as a local static variable?